core/string/translation.cpp

```diff
@@ -1300,6 +1333,275 @@ StringName TranslationServer::doc_translate_plural(const StringName &p_message,
+	}
+
+	if (pseudolocalization_double_vowels_enabled) {
+		message = double_vowels(message);
+	}
+
+	if (pseudolocalization_accents_enabled) {
+		message = replace_with_accented_string(message);
+	}
+
+	if (pseudolocalization_fake_bidi_enabled) {
+		message = wrap_with_fakebidi_characters(message);
+	}
+
+	StringName res = add_padding(message, length);
+	return res;
+}
+
+StringName TranslationServer::tool_pseudolocalize(const StringName &p_message) const {
+	String message = p_message;
+	message = double_vowels(message);
+	message = replace_with_accented_string(message);
+	StringName res = "[!!! " + message + " !!!]";
+	return res;
+}
+
+String TranslationServer::get_override_string(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				(message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += message[i];
+			res += message[i + 1];
+			i++;
+			continue;
+		}
+		res += '*';
+	}
+	return res;
+}
+String TranslationServer::double_vowels(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				(message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += message[i];
+			res += message[i + 1];
+			i++;
+			continue;
+		}
+		res += message[i];
+		if (message[i] == 'a' || message[i] == 'e' || message[i] == 'i' || message[i] == 'o' || message[i] == 'u' ||
+				message[i] == 'A' || message[i] == 'E' || message[i] == 'I' || message[i] == 'O' || message[i] == 'U') {
+			res += message[i];
+		}
+	}
+	return res;
+};
+
+String TranslationServer::replace_with_accented_string(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				(message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += message[i];
+			res += message[i + 1];
+			i++;
+			continue;
+		}
+		const char32_t *accented = get_accented_version(message[i]);
+		if (accented) {
+			res += accented;
+		} else {
+			res += message[i];
+		}
+	}
+	return res;
+}
+
+String TranslationServer::wrap_with_fakebidi_characters(String &message) const {
+	String res = "";
+	char32_t fakebidiprefix = U'\u202e';
+	char32_t fakebidisuffix = U'\u202c';
+	res += fakebidiprefix;
+	//the fake bidi unicode gets popped at every newline so pushing it back at every newline.
```

Comment:
This is the preferred comment style (for new code at least), space after `//`, starts with a capital letter, ends with a dot (i.e. a normal sentence).
```suggestion
	// The fake bidi unicode gets popped at every newline so pushing it back at every newline.
```

Applies to other comments in this PR.